appender.go

```diff
@@ -132,149 +150,329 @@ func (a *Appender) AppendRow(args ...driver.Value) error {
+		err := c.getChildVectors(vector)
+		if err != nil {
+			return err
+		}
 	}
 
 	a.chunks = append(a.chunks, dataChunk)
 	return nil
 }
 
+func setNull(columnInfo *ColumnInfo, rowIdx C.idx_t) {
+	C.duckdb_vector_ensure_validity_writable(columnInfo.vector)
+	mask := C.duckdb_vector_get_validity(columnInfo.vector)
+	C.duckdb_validity_set_row_invalid(mask, rowIdx)
+}
+
+func setPrimitive[T any](a *Appender, columnInfo *ColumnInfo, rowIdx C.idx_t, value T) {
+	ptr := C.duckdb_vector_get_data(columnInfo.vector)
+	xs := (*[1 << 31]T)(ptr)
+	xs[rowIdx] = value
+}
+
+func setVarchar(a *Appender, columnInfo *ColumnInfo, rowIdx C.idx_t, value string) {
+	str := C.CString(value)
+	C.duckdb_vector_assign_string_element(columnInfo.vector, rowIdx, str)
+	C.free(unsafe.Pointer(str))
+}
+
+func setTime(a *Appender, columnInfo *ColumnInfo, rowIdx C.idx_t, value driver.Value) {
+	var dt C.duckdb_timestamp
+	dt.micros = C.int64_t(value.(time.Time).UTC().UnixMicro())
+	setPrimitive[C.duckdb_timestamp](a, columnInfo, rowIdx, dt)
+}
+
+func setList(a *Appender, columnInfo *ColumnInfo, rowIdx C.idx_t, value driver.Value) {
+	refVal := reflect.ValueOf(value)
+
+	childColumnInfo := columnInfo.columnInfos[0]
+
+	childVectorSize := C.duckdb_list_vector_get_size(columnInfo.vector)
+
+	if refVal.IsNil() {
+		setNull(columnInfo, a.currentChunkSize)
+		childVectorSize += 1
+	}
+
+	// Convert the refVal to []interface{} to be able to iterate over it
+	interfaceSlice := make([]interface{}, refVal.Len())
+	for i := 0; i < refVal.Len(); i++ {
+		interfaceSlice[i] = refVal.Index(i).Interface()
+	}
+
+	// Set the offset of the list vector using the current size of the child vector.
+	// The duckdb list vector contains for each row two values, the offset and the length,
+	// which are located at rowIdx*2 and rowIdx*2+1 respectively.
+	setPrimitive[uint64](a, columnInfo, rowIdx*2, uint64(childVectorSize))
+
+	// Set the length of the list vector
+	setPrimitive[uint64](a, columnInfo, rowIdx*2+1, uint64(refVal.Len()))
+
+	C.duckdb_list_vector_set_size(columnInfo.vector, C.idx_t(refVal.Len())+childVectorSize)
+	C.duckdb_list_vector_reserve(columnInfo.vector, C.idx_t(refVal.Len())+childVectorSize)
+
+	// Insert the values into the child vector
+	for i, e := range interfaceSlice {
+		childVectorRow := C.idx_t(i) + childVectorSize
+		childColumnInfo.function(a, &childColumnInfo, childVectorRow, e)
+	}
+}
+
+func setStruct(a *Appender, columnInfo *ColumnInfo, rowIdx C.idx_t, value driver.Value) {
+	refVal := reflect.ValueOf(value)
+	structType := refVal.Type()
+
+	if value == nil {
+		setNull(columnInfo, a.currentChunkSize)
+	}
+
+	for i := 0; i < structType.NumField(); i++ {
+		childColumnInfo := columnInfo.columnInfos[i]
+		childColumnInfo.function(a, &childColumnInfo, rowIdx, refVal.Field(i).Interface())
+	}
+}
+
 // appendRowArray loads a row of values into the appender. The values are provided as an array.
 func (a *Appender) appendRowArray(args []driver.Value) error {
```

Comment:
I don't think this function has to return an `error`... as of this implementation. But what happens if the user passes a value that does not match the table/schema of the first row? We need to return an actual error here, and we need to return that error if our callbacks receive a value that does not match. Can we also get a test for this?

Reply:
Currently, if the user passes a value that doesn't match the table, `C.duckdb_append_data_chunk` throws an error.  The problem is that it only happens after everything has already been added into the data chunks and is only thrown once the user flushes or closes the appender. However, on the flip side, the only list of types we have here is this: `[]C.duckdb_logical_type`, which is not directly convertible to a go `Type`.  So either we would have to initialize and store an array of `Type's` to cross check, or convert the value to `C.duckdb_logical_type`.  What do you think?

Reply:
Also we only have access to the first row, which could still be different to schema, so we could throw an error for any row that doesn't match the first row, but it still could error out in `C.duckdb_append_data_chunk` if it doesn't match the schema